src/transfer-manager.ts

```diff
@@ -0,0 +1,349 @@
+/*!
+ * Copyright 2022 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Bucket, UploadOptions, UploadResponse} from './bucket';
+import {
+  DownloadCallback,
+  DownloadOptions,
+  DownloadResponse,
+  File,
+} from './file';
+import * as pLimit from 'p-limit';
+import {Metadata} from './nodejs-common';
+import * as path from 'path';
+import * as extend from 'extend';
+import {promises as fsp} from 'fs';
+
+const DEFAULT_PARALLEL_UPLOAD_LIMIT = 2;
+const DEFAULT_PARALLEL_DOWNLOAD_LIMIT = 2;
+const DEFAULT_PARALLEL_LARGE_FILE_DOWNLOAD_LIMIT = 2;
+const LARGE_FILE_SIZE_THRESHOLD = 256 * 1024 * 1024;
+const LARGE_FILE_DEFAULT_CHUNK_SIZE = 10 * 1024 * 1024;
+const EMPTY_REGEX = '(?:)';
+export interface UploadMultiOptions {
+  concurrencyLimit?: number;
+  skipIfExists?: boolean;
+  prefix?: string;
+  passthroughOptions?: UploadOptions;
+}
+
+export interface DownloadMultiOptions {
+  concurrencyLimit?: number;
+  prefix?: string;
+  stripPrefix?: string;
+  passthroughOptions?: DownloadOptions;
+}
+
+export interface LargeFileDownloadOptions {
+  concurrencyLimit?: number;
+  chunkSizeBytes?: number;
+}
+
+export interface UploadMultiCallback {
+  (err: Error | null, files?: File[], metadata?: Metadata[]): void;
+}
+
+export interface DownloadMultiCallback {
+  (err: Error | null, contents?: Buffer[]): void;
+}
+
+/**
+ * Create a TransferManager object to perform parallel transfer operations on a Cloud Storage bucket.
+ *
+ * @class
+ * @hideconstructor
+ *
+ * @param {Bucket} bucket A {@link Bucket} instance
+ */
+export class TransferManager {
+  bucket: Bucket;
+  constructor(bucket: Bucket) {
```

Comment:
Does Bucket pass through all the options from Storage (e.g. crc32c )?

Reply:
`Bucket` has a reference to `Storage` and a bucket instance cannot be created without providing it an instance of storage.